File: backend-legacy-python/app/vpn/protocols/xicmp.py
"""XICMP Tunnel — emergency fallback: proxy traffic over ICMP ping packets.

Even slower than XDNS but ICMP is almost never blocked by firewalls or DPI.
Uses xray Finalmask XICMP with mKCP transport wrapped in ICMP.

WARNING: LAST RESORT — extremely slow, unreliable on some networks.
Only enable when DNS tunnel also fails.
"""

from app.config import get_settings

from .base import ClientLink, Protocol, ServerConfig, UserCredentials, XrayInbound


class Xicmp(Protocol):
    def __init__(self) -> None:
        s = get_settings()
        self._enabled = s.xicmp_enabled

    @property
    def name(self) -> str:
        return "xicmp"

    @property
    def display_name(self) -> str:
        return "ICMP Tunnel (Emergency)"

    @property
    def enabled(self) -> bool:
        return self._enabled

    def xray_inbounds(
        self, users: list[UserCredentials], short_ids: list[str]
    ) -> list[XrayInbound]:
        if not self.enabled:
            return []
        return [
            XrayInbound(
                tag="xicmp-in",
                port=0,  # ICMP is portless — raw socket
                protocol="dokodemo-door",
                settings={
                    "network": "tcp,udp",
                    "followRedirect": False,
                },
                stream_settings={
                    "network": "mkcp",
                    "kcpSettings": {
                        "header": {"type": "none"},
                        "seed": "",
                    },
                    "finalmask": {"type": "xicmp"},
                },
                sniffing={"enabled": True, "destOverride": ["http", "tls"]},
            )
        ]

    def xray_outbounds(self) -> list[dict]:
        if not self.enabled:
            return []
        return [
            {
                "protocol": "freedom",
                "tag": "xicmp-out",
                "streamSettings": {
                    "finalmask": {"type": "xicmp"},
                },
            }
        ]

    def xray_routing_rules(self) -> list[dict]:
        if not self.enabled:
            return []
        return [
            {
                "type": "field",
                "inboundTag": ["xicmp-in"],
                "outboundTag": "xicmp-out",
            }
        ]

    def client_links(
        self, user: UserCredentials, servers: list[ServerConfig]
    ) -> list[ClientLink]:
        """XICMP tunnel links — xray-native only."""
        if not self.enabled:
            return []
        links: list[ClientLink] = []
        for srv in servers:
            remark = f"{srv.flag} {srv.name} ICMP-Tunnel"
            uri = (
                f"xicmp://{user.uuid}@{srv.host}"
                f"?transport=mkcp"
                f"&security=none"
                f"#{remark}"
            )
            links.append(
                ClientLink(
                    uri=uri,
                    protocol="xicmp",
                    transport="icmp",
                    server_key=srv.key,
                    remark=remark,
                )
            )
        return links

    def singbox_outbound(
        self, tag: str, server: ServerConfig, user: UserCredentials, **opts
    ) -> dict | None:
        # XICMP is NOT supported by sing-box — clients must use xray-based config
        return None

File: backend/crates/chameleon-vpn/src/protocols/xicmp.rs
//! XICMP Tunnel — emergency fallback: proxy traffic over ICMP ping packets.

use serde_json::json;
use chameleon_config::Settings;
use super::types::*;

pub struct Xicmp {
    is_enabled: bool,
}

impl Xicmp {
    pub fn new(s: &Settings) -> Self {
        Self { is_enabled: s.xicmp_enabled }
    }
}

impl Protocol for Xicmp {
    fn name(&self) -> &str { "xicmp" }
    fn display_name(&self) -> &str { "ICMP Tunnel (Emergency)" }
    fn enabled(&self) -> bool { self.is_enabled }

    fn xray_inbounds(&self, _: &[UserCredentials], _: &[String]) -> Vec<XrayInbound> {
        if !self.enabled() { return vec![]; }
        vec![XrayInbound {
            tag: "xicmp-in".into(), port: 0, protocol: "dokodemo-door".into(),
            settings: json!({"network": "tcp,udp", "followRedirect": false}),
            stream_settings: json!({"network": "mkcp", "kcpSettings": {"header": {"type": "none"}, "seed": ""}, "finalmask": {"type": "xicmp"}}),
            sniffing: json!({"enabled": true, "destOverride": ["http", "tls"]}),
            listen: "0.0.0.0".into(),
        }]
    }

    fn xray_outbounds(&self) -> Vec<serde_json::Value> {
        if !self.enabled() { return vec![]; }
        vec![json!({"protocol": "freedom", "tag": "xicmp-out", "streamSettings": {"finalmask": {"type": "xicmp"}}})]
    }

    fn xray_routing_rules(&self) -> Vec<serde_json::Value> {
        if !self.enabled() { return vec![]; }
        vec![json!({"type": "field", "inboundTag": ["xicmp-in"], "outboundTag": "xicmp-out"})]
    }

    fn client_links(&self, user: &UserCredentials, servers: &[ServerConfig]) -> Vec<ClientLink> {
        if !self.enabled() { return vec![]; }
        servers.iter().map(|srv| {
            let remark = format!("{} {} ICMP-Tunnel", srv.flag, srv.name);
            let uri = format!("xicmp://{}@{}?transport=mkcp&security=none#{}", user.uuid, srv.host, remark);
            ClientLink { uri, protocol: "xicmp".into(), transport: "icmp".into(), server_key: srv.key.clone(), remark, is_relay: false }
        }).collect()
    }

    fn singbox_outbound(&self, _: &str, _: &ServerConfig, _: &UserCredentials, _: &OutboundOpts) -> Option<serde_json::Value> {
        None // Not supported by sing-box
    }
}
